lms/services/digest.py

```diff
@@ -0,0 +1,264 @@
+from dataclasses import dataclass, field
+from typing import List, Optional
+
+from sqlalchemy import func, select
+
+from lms.models import (
+    Assignment,
+    AssignmentGrouping,
+    AssignmentMembership,
+    Course,
+    Grouping,
+    LTIRole,
+    User,
+)
+from lms.services.h_api import HAPI
+from lms.services.mailchimp import EmailRecipient, EmailSender, MailchimpService
+
+
+class DigestService:
+    def __init__(
+        self, db, helper, h_api, mailchimp_service, sender
+    ):  # pylint:disable=too-many-arguments
+        self.db = db
+        self.helper = helper
+        self.h_api = h_api
+        self.mailchimp_service = mailchimp_service
+        self.sender = sender
+
+    def send_instructor_email_digests(
+        self, h_userids, since, until, override_to_email=None
+    ):
+        annotations = self.h_api.get_annotations(h_userids, since, until)
+        unified_courses = self.helper.unified_courses(annotations)
+
+        for h_userid in h_userids:
+            unified_user = UnifiedUser.make(self.db, h_userid)
+            digest = self.helper.instructor_digest(unified_user, unified_courses)
+
+            if not digest["num_annotations"]:
+                continue
+
+            if override_to_email is None:
+                to_email = unified_user.email
+            else:
+                to_email = override_to_email
+
+            self.mailchimp_service.send_template(
+                "instructor-email-digest",
+                self.sender,
+                recipient=EmailRecipient(to_email, unified_user.display_name),
+                template_vars=digest,
+            )
+
+
+@dataclass
+class UnifiedUser:
+    """All User's for a given h_userid, unified across all ApplicationInstance's."""
+
+    users: List[User] = field(default_factory=list)
+
+    def __post_init__(self):
+        for user in self.users:
+            assert user.h_userid == self.h_userid
+
+    @property
+    def h_userid(self) -> str:
+        """Return this user's h_userid."""
+        return self.users[0].h_userid
+
+    @property
+    def user_ids(self) -> List[int]:
+        """Return the User.id's of all the users with self.h_userid."""
+        return [user.id for user in self.users]
+
+    @property
+    def email(self) -> Optional[str]:
+        """Return an email address for this user."""
+        for user in self.users:
+            if user.email:
+                return user.email
+
+        return None
+
+    @property
+    def display_name(self) -> Optional[str]:
+        """Return a display name for this user."""
+        for user in self.users:
+            if user.display_name:
+                return user.display_name
+
+        return None
+
+    @classmethod
+    def make(cls, db, h_userid):
+        """Return a UnifiedUser for the given h_userid."""
+        return cls(
+            db.scalars(
+                select(User).filter_by(h_userid=h_userid).order_by(User.updated.desc())
+            ).all()
+        )
```

Comment:
This is going to do a separate DB query for each user.

We might be able to make this more efficient by doing one big query for all users but that would be harder to write and might be a premature optimisation and make the code harder to read.

It also might query for the _same_ user multiple times if the code calls it repeatedly with the same `h_userid` and I don't want to add `@lru_cache` to it because I can't clear the cache between unit tests (this seems to be an issue with `@lru_cache` and `@classmethod`).

This is probably going to be fine, so maybe we can monitor it and optimise if necessary.